packages/plugin-inline/Inline.ts

```diff
@@ -28,35 +38,90 @@ export class Inline extends JWPlugin {
-                inline.formats.splice(index, 1);
+        if (range.isCollapsed()) {
+            if (!this.formatCache) {
+                this.formatCache = this.getCurrentFormats(range);
+            }
+            const index = this.formatCache.findIndex(f => f instanceof FormatClass);
+            if (index !== -1) {
+                this.formatCache.splice(index, 1);
+            } else {
+                this.formatCache.push(new FormatClass());
             }
         } else {
-            // If there is at least one char in the range without the format
-            // `FormatClass`, set the format for all nodes.
-            for (const inline of selectedInlines) {
-                if (!inline.formats.find(f => f instanceof FormatClass)) {
-                    new FormatClass().applyTo(inline);
+            const selectedInlines = range.selectedNodes(InlineNode);
+
+            // If every char in the range has the format `FormatClass`, remove
+            // the format for all of them.
+            const allHaveFormat = selectedInlines.every(inline => {
+                return !!inline.formats.find(f => f instanceof FormatClass);
+            });
+            if (allHaveFormat) {
+                for (const inline of selectedInlines) {
+                    const index = inline.formats.findIndex(f => f instanceof FormatClass);
+                    // Apply the attributes of the format we're about to remove to
+                    // the inline itself.
+                    const matchingFormat = inline.formats[index];
+                    for (const key of Object.keys(matchingFormat.attributes)) {
+                        inline.attributes[key] = matchingFormat.attributes[key];
+                    }
+                    // Remove the format.
+                    inline.formats.splice(index, 1);
                 }
+            } else {
+                // If there is at least one char in the range without the format
+                // `FormatClass`, set the format for all nodes.
+                for (const inline of selectedInlines) {
+                    if (!inline.formats.find(f => f instanceof FormatClass)) {
+                        new FormatClass().applyTo(inline);
+                    }
+                }
+            }
+        }
+    }
+    isAllFormat(
+        FormatClass: Constructor<Format>,
+        range = this.editor.vDocument.selection.range,
+    ): boolean {
+        if (range.isCollapsed()) {
+            if (!this.formatCache) {
+                this.formatCache = this.getCurrentFormats(range);
             }
+            return !!this.formatCache.find(format => format instanceof FormatClass);
+        } else {
+            const selectedInlines = range.selectedNodes(InlineNode);
+            return (
+                selectedInlines.length &&
+                selectedInlines.every(
+                    char => !!char.formats.find(format => format instanceof FormatClass),
+                )
+            );
```

Comment:
I don't like this function. It looks very ad-hoc to me. I'd like to replace it by a function that returns all formats that are active at the range. That means all formats that are shared by all the selected nodes plus the formats in currently in `formatCache`. Then you'd just have to check if your format is in there.